ui: Cluster Viz Treemap

mrtracy opened the pull request:
![treemapgif](https://user-images.githubusercontent.com/6969858/33854555-1d2df346-de90-11e7-9b0b-0af28d10a9b3.gif)

4 commits which create in the initial factoring for a switchable treemap/map view.  Nodes and aggregated localities now have a position on both the geographic map and a hierarchical treemap layout. The user can toggle between the two layouts with a smooth animation.

Includes new dependency of react-motion to enable the animation.

Commit:
ui: Move AccumulateProps to lifecycle methods

Addresses a previously identified TODO.

Release Notes: None

Commit:
ui: Isolate node simulation into its own component

Previously, the CurrentNodes component had been responsible for
accumulating node histories. This responsibility has been refactored to
a dedicated NodeSimulator component.

Release Notes: None

Commit:
ui: Locality Treemap Layout

Large change that adds several abilities to the cluster visualization:

+ Cluster view now computes Localities based on available nodes. This
creates a heirarchy of nodes that can be used to aggregate nodes with
common locality tiers.
+ Adds "TreeMap" view of cluster, which displays all nodes on the cluster
organized into a series of nested rectangles by locality.
+ Map view has been modified to display only the top locality level,
rather than displaying all nodes.
+ Clicking on a locality will transition between the treemap and map
view with a smooth animation.

Also included is some refactoring, primarily to clarify the difference
between a computed locality and the NodeView component, which was
previously named "locality". Also moves some components into different
files in order to improve file cohesion.

Release Notes: None

Commit:
ui: Add react-motion, upgrade react types

Need to add react-motion, but the types for this library
(@types/react-motion) require an upgrade to our current react types
(@types/react), which appear to be somewhat out of date.  Three components
to this:

- Upgrade @types/react
- Add an explicit *resolve* in package.json for @types/react, which
forces there to be only one version of the library in use by all
packages. This is necessary because @types/react apparently defines some
global variables, and these can conflict if there are two versions of
the library.
- Add react-motion and @types/react-motion

While upgrading, the enzyme package got into a weird state and I could not
get yarn to stabilize on the previous setting in the lockfile; it
appears that the version we were using is simply no longer available,
which I don't understand. I upgraded it.

Release Notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20632)
<!-- Reviewable:end -->

dianasaur323 commented:
Nice animation! Pretty cool :)

couchand commented:
Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx, line 60 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BFf6tWaJtn75c72qV:-L0BFf6tWaJtn75c72qW:b-egn8nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx#L60)):*
> ```XML
>       let centroid: [number, number] = [0, 0];
>       this.children.forEach(c => centroid = vector.add(centroid, c.longLat()));
>       return vector.mult(centroid, 1 / this.children.length);
> ```

I think it would be nice to precalculate this, since it's a little expensive.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx, line 63 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BFYfC66FdiD9NDIUL:-L0BFYfC66FdiD9NDIUM:bnx32q2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx#L63)):*
> ```XML
>     }
> 
>     return [0, 0];
> ```

good old null island!

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx, line 93 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BGs-S5EvxSv6Fo_6n:-L0BGs-S5EvxSv6Fo_6o:b-tqsby8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx#L93)):*
> ```XML
>   addNode(node: SimulatedNodeStatus) {
>     const locality = this.getOrCreateTreeNode(node.tiers());
>     locality.children.push(new LocalityTreeNode(node));
> ```

I think this is just a little confusing, that `getOrCreateTreeNode` creates the parent, not the node itself.  Roughly the same forces require the duplication of the `.children.push` calls here and at the end of that function.

This works as is, we could probably stand to leave it and come clean it up later.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx, line 101 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BHNEmD9prHAWXiA0j:-L0BHNEmD9prHAWXiA0k:bvv7xro) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx#L101)):*
> ```XML
>     treeMap.ratio(1);
>     treeMap.padding(30);
>     treeMap.value(n => n.data instanceof SimulatedNodeStatus ? 1 : 0);
> ```

maybe use `n.isNode()` instead?

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 34 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BJC08pkpRw8hM9QyM:-L0BJC08pkpRw8hM9QyN:br711an) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L34)):*
> ```XML
>     let childNodes: JSX.Element[];
>     if (showChildren && data.children) {
>       /* tslint:disable-next-line:no-use-before-declare */
> ```

what's this lint escape for?

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/nodeSimulator.tsx, line 1 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BHbwzB-medmjU_Cvn:-L0BHbwzB-medmjU_Cvo:b-irete6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/nodeSimulator.tsx#L1)):*
> ```XML
> // Copyright 2017 The Cockroach Authors.
> ```

Is this whole file to be eventually deleted and replaced by API calls?  Maybe it would be nice to have a big comment to that effect at the top, or clarify which pieces are expected to be integrated elsewhere?

---

*[pkg/ui/src/views/shared/components/dropdown/index.tsx, line 54 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BE8LAUSCdD3kRXrma:-L0BE8LBndf94TuWLpqn:b54dy73) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/src/views/shared/components/dropdown/index.tsx#L54)):*
> ```XML
>     return <div className={className}>
>       {/* TODO (maxlang): consider moving arrows outside the dropdown component */}
>       <button
> ```

Are these changes related?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx, line 60 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BFf6tWaJtn75c72qV:-L0CuoBvUU9IZbKjbM1N:b3ches0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx#L60)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think it would be nice to precalculate this, since it's a little expensive.
</blockquote></details>

Added a TODO to memoize it, there's gonna be some more refactoring in here before i'm finished.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx, line 63 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BFYfC66FdiD9NDIUL:-L0Cud6LE4hfQ0l9GAKz:b-ngj3nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx#L63)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

good old null island!
</blockquote></details>

Eh, it really shouldn't happen.  I've taken it out.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx, line 101 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BHNEmD9prHAWXiA0j:-L0CuNURDz8SQZ3w8bh-:boh7v9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/localityCollection.tsx#L101)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

maybe use `n.isNode()` instead?
</blockquote></details>

I added it for this very reason! Fixed.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 34 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BJC08pkpRw8hM9QyM:-L0Cqu3KTEpBnHzzp8iS:b-pnla1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L34)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

what's this lint escape for?
</blockquote></details>

ModalTreeNodeView uses LocalityBox, LocalityBox uses ModalTreeNodeView. It's a recursive structure, i'm not sure how to define it to avoid falling afoul of this rule, this was easier.

I have thought of a way to implement this without the recursion, i'm implementing it as part of the next commit set.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/nodeSimulator.tsx, line 1 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BHbwzB-medmjU_Cvn:-L0CxfMC6MpNeynX7tnQ:b-eopdcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/ccl/src/views/clusterviz/containers/map/nodeSimulator.tsx#L1)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Is this whole file to be eventually deleted and replaced by API calls?  Maybe it would be nice to have a big comment to that effect at the top, or clarify which pieces are expected to be integrated elsewhere?
</blockquote></details>

Added numerous comments that this should be removed before release.

---

*[pkg/ui/src/views/shared/components/dropdown/index.tsx, line 54 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632#-L0BE8LAUSCdD3kRXrma:-L0CpX6gBe-W-tSFSqlc:b-gsqlfv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a8407fbe115da580d2de2429a49f9b42f05680f/pkg/ui/src/views/shared/components/dropdown/index.tsx#L54)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Are these changes related?
</blockquote></details>

Yeah it was necessary for an updated react type (span shouldn't have allowed the 'disabled' attribute), but apparently this was already addressed as part of your classNames refactor.  I'll just revert this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20632)*
<!-- Sent from Reviewable.io -->
